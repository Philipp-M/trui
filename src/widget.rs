mod border;
mod core;
mod events;
mod linear_layout;
mod text;

pub(crate) use self::core::WidgetState;
pub use self::core::{
    AnyWidget, ChangeFlags, CxState, Event, EventCx, LayoutCx, PaintCx, Pod, Point,
    StyleableWidget, Widget,
};
pub use border::Border;
pub use events::{OnClick, OnHover, OnHoverLost, StyleOnHover};
pub use linear_layout::LinearLayout;
pub use ratatui::layout::Rect;
pub use text::Text;
