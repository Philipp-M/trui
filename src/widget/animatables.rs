use std::{any::Any, ops::DerefMut, time::Duration};

use super::{ChangeFlags, LifeCycleCx};

pub trait AnimatableElement<V>: 'static + AnyAnimatableElement<V> {
    fn animate(&mut self, cx: &mut LifeCycleCx) -> &V;
}

pub trait AnyAnimatableElement<V> {
    fn as_any(&self) -> &dyn Any;

    fn as_any_mut(&mut self) -> &mut dyn Any;

    fn type_name(&self) -> &'static str;
}

impl<V, A: AnimatableElement<V>> AnyAnimatableElement<V> for A {
    fn as_any(&self) -> &dyn Any {
        self
    }

    fn as_any_mut(&mut self) -> &mut dyn Any {
        self
    }

    fn type_name(&self) -> &'static str {
        std::any::type_name::<Self>()
    }
}

impl<V: 'static> AnimatableElement<V> for Box<dyn AnimatableElement<V>> {
    fn animate(&mut self, cx: &mut LifeCycleCx) -> &V {
        self.deref_mut().animate(cx)
    }
}

macro_rules! impl_animatable_for_primitive {
    ($type: ty) => {
        impl AnimatableElement<$type> for $type {
            fn animate(&mut self, _cx: &mut LifeCycleCx) -> &$type {
                self
            }
        }
    };
}

// All builtin number types
impl_animatable_for_primitive!(i8);
impl_animatable_for_primitive!(u8);
impl_animatable_for_primitive!(i16);
impl_animatable_for_primitive!(u16);
impl_animatable_for_primitive!(i32);
impl_animatable_for_primitive!(u32);
impl_animatable_for_primitive!(i64);
impl_animatable_for_primitive!(u64);
impl_animatable_for_primitive!(i128);
impl_animatable_for_primitive!(u128);
impl_animatable_for_primitive!(isize);
impl_animatable_for_primitive!(usize);

impl_animatable_for_primitive!(f32);
impl_animatable_for_primitive!(f64);

#[derive(Clone, Debug)]
pub struct LowPassIIR<AT, V> {
    pub(crate) target: AT,
    decay: f64,
    value: Option<V>,
}

impl<AT> LowPassIIR<AT, f64> {
    pub(crate) fn new(target: AT, decay: f64) -> Self {
        LowPassIIR {
            target,
            decay,
            value: None,
        }
    }

    pub(crate) fn set_decay(&mut self, decay: f64) -> ChangeFlags {
        if self.decay != decay {
            self.decay = decay;
            ChangeFlags::ANIMATION
        } else {
            ChangeFlags::empty()
        }
    }
}

impl<AT: AnimatableElement<f64>> AnimatableElement<f64> for LowPassIIR<AT, f64> {
    fn animate(&mut self, cx: &mut LifeCycleCx) -> &f64 {
        let target_value = self.target.animate(cx);
        if let Some(value) = &mut self.value {
            if (*target_value - *value).abs() > 0.0001 {
                let delta_time = cx.time_since_last_render_request().as_secs_f64() * 100.0; // could be a different factor, and maybe more precisely a frequency based cutoff or something like that
                let time_adjusted_decay =
                    1.0 - ((1.0 - self.decay.clamp(0.0, 1.0)).powf(delta_time));
                *value += time_adjusted_decay * (*target_value - *value);
                cx.request_animation_update();
            } else if *value != *target_value {
                *value = *target_value;
                cx.request_animation_update();
            }
        } else {
            self.value = Some(*target_value);
        }
        self.value.as_ref().unwrap()
    }
}

#[derive(Clone, Debug)]
pub struct Lerp<T, R> {
    pub(crate) tweenable: T,
    pub(crate) ratio: R,
}

impl<T, R> Lerp<T, R> {
    pub(crate) fn new(tweenable: T, ratio: R) -> Self {
        Lerp { tweenable, ratio }
    }
}

impl<V, T: TweenableElement<V>, R: AnimatableElement<f64>> AnimatableElement<V> for Lerp<T, R> {
    fn animate(&mut self, cx: &mut LifeCycleCx) -> &V {
        let ratio = self.ratio.animate(cx);
        self.tweenable.interpolate(cx, *ratio)
    }
}

// TODO Duration could also be animated, but I'm not sure it's worth the complexity (vs benefit)...
#[derive(Clone, Debug)]
pub struct PlayTween<PS, TW> {
    pub(crate) play_speed: PS,
    current_time: Duration,
    pub(crate) tweenable: TW,
}

impl<PS, TW> PlayTween<PS, TW> {
    pub(crate) fn new(play_speed: PS, tweenable: TW) -> Self {
        PlayTween {
            play_speed,
            tweenable,
            current_time: Duration::ZERO,
        }
    }
}

impl<V: 'static, PS: AnimatableElement<f64>, TW: TweenableElement<V>> AnimatableElement<V>
    for PlayTween<PS, TW>
{
    fn animate(&mut self, cx: &mut LifeCycleCx) -> &V {
        let play_speed = self.play_speed.animate(cx);
        let duration_as_secs = self.tweenable.duration().as_secs_f64();
        let current_time_as_secs = self.current_time.as_secs_f64();
        let new_time = (current_time_as_secs
            + *play_speed * cx.time_since_last_render_request().as_secs_f64())
        .clamp(0.0, duration_as_secs);
        // avoid division by zero
        let ratio = if duration_as_secs == 0.0 {
            0.0
        } else {
            current_time_as_secs / duration_as_secs
        };

        if !self.tweenable.duration().is_zero()
            && ((*play_speed > 0.0 && self.current_time != self.tweenable.duration())
                || (*play_speed < 0.0 && self.current_time != Duration::ZERO))
        {
            self.current_time = Duration::from_secs_f64(new_time);
            cx.request_animation_update();
        }
        self.tweenable.interpolate(cx, ratio)
    }
}

// ---------------------------------- TWEENABLE

pub trait TweenableElement<V>: 'static + AnyTweenableElement<V> {
    fn interpolate(&mut self, cx: &mut LifeCycleCx, ratio: f64) -> &V;

    // TODO &mut?
    // Could also be collecting a default from some kind of context (LifeCycleCx?)
    /// Default duration is 1 second
    fn duration(&mut self) -> Duration {
        Duration::from_secs(1)
    }
}

pub trait AnyTweenableElement<V> {
    fn as_any(&self) -> &dyn Any;

    fn as_any_mut(&mut self) -> &mut dyn Any;

    fn type_name(&self) -> &'static str;
}

impl<V, A: TweenableElement<V>> AnyTweenableElement<V> for A {
    fn as_any(&self) -> &dyn Any {
        self
    }

    fn as_any_mut(&mut self) -> &mut dyn Any {
        self
    }

    fn type_name(&self) -> &'static str {
        std::any::type_name::<Self>()
    }
}

impl<V: 'static> TweenableElement<V> for Box<dyn TweenableElement<V>> {
    fn interpolate(&mut self, cx: &mut LifeCycleCx, ratio: f64) -> &V {
        self.deref_mut().interpolate(cx, ratio)
    }

    // TODO &mut?
    // Could also be collecting a default from some kind of context (LifeCycleCx?)
    /// Default duration is 1 second
    fn duration(&mut self) -> Duration {
        self.deref_mut().duration()
    }
}

pub struct TweenableRange<A, V> {
    pub(crate) start: A,
    pub(crate) end: A,
    value: V,
}

impl<A> TweenableRange<A, f64> {
    pub(crate) fn new(start: A, end: A) -> Self {
        Self {
            start,
            end,
            value: 0.0,
        }
    }
}

impl<A: AnimatableElement<f64>> TweenableElement<f64> for TweenableRange<A, f64> {
    fn interpolate(&mut self, cx: &mut LifeCycleCx, ratio: f64) -> &f64 {
        let start = self.start.animate(cx);
        let end = self.end.animate(cx);

        self.value = *start * (1.0 - ratio) + *end * ratio;
        &mut self.value
    }
}

pub struct Map<I, V, VO> {
    pub(crate) input: I,
    output: Option<VO>,
    f: fn(&V) -> VO,
}

impl<I, V, VO> Map<I, V, VO> {
    pub fn new(input: I, f: fn(&V) -> VO) -> Self {
        Map {
            input,
            output: None,
            f,
        }
    }

    pub fn update_f(&mut self, f: fn(&V) -> VO) -> ChangeFlags {
        if self.f != f {
            self.f = f;
            ChangeFlags::ANIMATION
        } else {
            ChangeFlags::empty()
        }
    }
}

impl<VO: 'static, V: 'static, I: TweenableElement<V>> TweenableElement<VO> for Map<I, V, VO> {
    fn interpolate(&mut self, cx: &mut LifeCycleCx, ratio: f64) -> &VO {
        let input = self.input.interpolate(cx, ratio);
        self.output = Some((self.f)(input));
        self.output.as_ref().unwrap()
    }

    fn duration(&mut self) -> Duration {
        self.input.duration()
    }
}

/// Overrides the duration of any tweenable it composes
pub struct WithDuration<T> {
    pub(crate) tweenable: T,
    pub(crate) duration: Duration,
}

impl<T> WithDuration<T> {
    pub(crate) fn new(tweenable: T, duration: Duration) -> Self {
        Self {
            tweenable,
            duration,
        }
    }
}

impl<V, T: TweenableElement<V>> TweenableElement<V> for WithDuration<T> {
    fn interpolate(&mut self, cx: &mut LifeCycleCx, ratio: f64) -> &V {
        self.tweenable.interpolate(cx, ratio)
    }

    fn duration(&mut self) -> Duration {
        self.duration
    }
}

pub struct Sequence<V> {
    pub(crate) tweenables: Vec<Box<dyn TweenableElement<V>>>,
}

impl<V> Sequence<V> {
    pub fn new(tweenables: Vec<Box<dyn TweenableElement<V>>>) -> Self {
        Self { tweenables }
    }
}

impl<V: 'static> TweenableElement<V> for Sequence<V> {
    fn interpolate(&mut self, cx: &mut LifeCycleCx, mut ratio: f64) -> &V {
        let total_duration = self
            .tweenables
            .iter_mut()
            .fold(Duration::ZERO, |acc, tweenable| acc + tweenable.duration());
        let total_duration_f64 = total_duration.as_secs_f64();

        if self.tweenables.is_empty() {
            panic!("A Sequence should never be empty");
        }

        let mut duration_acc = Duration::ZERO;
        let mut target_ix = None;

        for (ix, tweenable) in self.tweenables.iter_mut().enumerate() {
            let tween_duration = tweenable.duration();
            let next_duration_acc = duration_acc + tween_duration;
            let start_ratio = duration_acc.as_secs_f64() / total_duration_f64;
            let end_ratio = next_duration_acc.as_secs_f64() / total_duration_f64;

            if ratio >= start_ratio && ratio < end_ratio {
                target_ix = Some(ix);
                ratio = (ratio - start_ratio) / (end_ratio - start_ratio);
                break;
            }

            duration_acc = next_duration_acc;
        }

        if let Some(ix) = target_ix {
            self.tweenables[ix].interpolate(cx, ratio)
        } else {
            self.tweenables.last_mut().unwrap().interpolate(cx, 1.0)
        }
    }

    fn duration(&mut self) -> Duration {
        self.tweenables
            .iter_mut()
            .fold(Duration::ZERO, |acc, tweenable| acc + tweenable.duration())
    }
}

pub mod ease {
    use crate::widget::LifeCycleCx;
    use std::time::Duration;

    use super::TweenableElement;

    macro_rules! ease_fn {
        ($type: ident, $ease_fn: expr) => {
            #[derive(Clone, Debug)]
            pub struct $type<T>(pub(crate) T);

            impl<V, T: TweenableElement<V>> TweenableElement<V> for $type<T> {
                fn interpolate(&mut self, cx: &mut LifeCycleCx, ratio: f64) -> &V {
                    #[allow(clippy::redundant_closure_call)]
                    self.0.interpolate(cx, $ease_fn(ratio))
                }

                fn duration(&mut self) -> Duration {
                    self.0.duration()
                }
            }
        };
    }

    ease_fn!(Reverse, |ratio| 1.0 - ratio);
    ease_fn!(QuadraticIn, |ratio| ratio * ratio);
    ease_fn!(QuadraticOut, |ratio: f64| -(ratio * (ratio - 2.0)));
    ease_fn!(QuadraticInOut, |ratio: f64| {
        if ratio < 0.5 {
            2.0 * ratio * ratio
        } else {
            (-2.0 * ratio * ratio) + (4.0 * ratio) - 1.0
        }
    });
    ease_fn!(ElasticInOut, |ratio: f64| {
        use std::f64::consts::TAU;
        (if ratio < 0.5 {
            0.5 * (13.0 * TAU * (2.0 * ratio)).sin() * 2.0_f64.powf(10.0 * ((2.0 * ratio) - 1.0))
        } else {
            0.5 * ((-13.0 * TAU * ((2.0 * ratio - 1.0) + 1.0)).sin()
                * 2.0_f64.powf(-10.0 * (2.0 * ratio - 1.0))
                + 2.0)
        })
        .clamp(0.0, 1.0)
    });
}
