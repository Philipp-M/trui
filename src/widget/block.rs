use super::{
    core::LayoutCx, core::PaintCx, BoxConstraints, ChangeFlags, Event, EventCx, Pod, Widget,
};
use crate::{
    geometry::{to_ratatui_rect, Point, Size},
    view::{BorderStyles, Borders},
};
use ratatui::{style::Style, symbols};

pub(crate) fn render_border(
    cx: &mut PaintCx,
    r: ratatui::layout::Rect,
    border_styles: &BorderStyles,
    style: Style,
) {
    use Borders as B; // unfortunately not possible to wildcard import since it's not an enum...
    if r.width == 0 || r.height == 0 {
        return;
    }
    let buf = cx.terminal.current_buffer_mut();

    let mut draw = |x, y, symbol, style| {
        if buf.area.x + x < buf.area.width && buf.area.y + y < buf.area.height {
            buf.get_mut(x, y).set_symbol(symbol).set_style(style);
        }
    };

    if r.width == 1 && r.height == 1 {
        let style = border_styles.style(B::ALL).patch(style);
        draw(r.x, r.y, symbols::DOT, style);
        return;
    }

    if r.width > 1 && border_styles.has_borders(B::TOP) {
        let style = border_styles.style(B::TOP).patch(style);
        let symbol = border_styles.symbols(B::TOP).horizontal;
        for x in r.x..(r.x + r.width) {
            draw(x, r.y, symbol, style);
        }
    }

    if r.width > 1 && border_styles.has_borders(B::BOTTOM) {
        let style = border_styles.style(B::BOTTOM).patch(style);
        let symbol = border_styles.symbols(B::BOTTOM).horizontal;
        for x in r.x..(r.x + r.width) {
            draw(x, r.y + r.height - 1, symbol, style);
        }
    }

    if r.height > 1 && border_styles.has_borders(B::LEFT) {
        let style = border_styles.style(B::LEFT).patch(style);
        let symbol = border_styles.symbols(B::LEFT).vertical;
        for y in r.y..(r.y + r.height) {
            draw(r.x, y, symbol, style);
        }
    }

    if r.height > 1 && border_styles.has_borders(B::RIGHT) {
        let style = border_styles.style(B::RIGHT).patch(style);
        let symbol = border_styles.symbols(B::RIGHT).vertical;
        for y in r.y..(r.y + r.height) {
            draw(r.x + r.width - 1, y, symbol, style);
        }
    }

    // corners
    if r.width > 1 && r.height > 1 {
        if border_styles.has_borders(B::LEFT | B::TOP) {
            let style = border_styles.style(B::LEFT | B::TOP).patch(style);
            let symbol = border_styles.symbols(B::LEFT | B::TOP).top_left;
            draw(r.x, r.y, symbol, style);
        }
        if border_styles.has_borders(B::LEFT | B::BOTTOM) {
            let style = border_styles.style(B::LEFT | B::BOTTOM).patch(style);
            let symbol = border_styles.symbols(B::LEFT | B::BOTTOM).bottom_left;
            draw(r.x, r.y + r.height - 1, symbol, style);
        }
        if border_styles.has_borders(B::RIGHT | B::BOTTOM) {
            let style = border_styles.style(B::RIGHT | B::BOTTOM).patch(style);
            let symbol = border_styles.symbols(B::RIGHT | B::BOTTOM).bottom_right;
            draw(r.x + r.width - 1, r.y + r.height - 1, symbol, style);
        }
        if border_styles.has_borders(B::RIGHT | B::TOP) {
            let style = border_styles.style(B::RIGHT | B::TOP).patch(style);
            let symbol = border_styles.symbols(B::RIGHT | B::TOP).top_right;
            draw(r.x + r.width - 1, r.y, symbol, style);
        }
    }
}

fn fill_block(cx: &mut PaintCx, r: ratatui::layout::Rect, style: Style) {
    let buf = cx.terminal.current_buffer_mut();

    for x in r.x..(buf.area.width.min(r.width + r.x)) {
        for y in r.y..(buf.area.height.min(r.height + r.y)) {
            buf.get_mut(x, y).set_style(style);
        }
    }
}

pub struct Block {
    pub(crate) content: Pod,
    border_styles: BorderStyles,
    style: Style,
    fill_with_bg: bool,
    inherit_style: bool,
}

impl Block {
    pub(crate) fn new(
        content: impl Widget,
        border_styles: BorderStyles,
        style: Style,
        inherit_style: bool,
        fill_with_bg: bool,
    ) -> Self {
        Block {
            content: Pod::new(content),
            fill_with_bg,
            style,
            inherit_style,
            border_styles,
        }
    }

    pub(crate) fn set_border_style(&mut self, border_style: &BorderStyles) -> ChangeFlags {
        if &self.border_styles != border_style {
            self.border_styles = border_style.clone();
            // TODO more sophisticated check for needed ChangeFlags (specifically layout)
            ChangeFlags::LAYOUT | ChangeFlags::PAINT
        } else {
            ChangeFlags::empty()
        }
    }

    pub(crate) fn set_fill_with_bg(&mut self, fill_with_bg: bool) -> ChangeFlags {
        if self.fill_with_bg != fill_with_bg {
            self.fill_with_bg = fill_with_bg;
            ChangeFlags::PAINT
        } else {
            ChangeFlags::empty()
        }
    }

    pub(crate) fn set_inherit_style(&mut self, inherit: bool) -> ChangeFlags {
        if self.inherit_style != inherit {
            self.inherit_style = inherit;
            ChangeFlags::PAINT
        } else {
            ChangeFlags::empty()
        }
    }
    pub(crate) fn set_style(&mut self, style: Style) -> ChangeFlags {
        if style != self.style {
            self.style = style;
            ChangeFlags::PAINT
        } else {
            ChangeFlags::empty()
        }
    }
}

impl Widget for Block {
    fn paint(&mut self, cx: &mut PaintCx) {
        let style = self.style.patch(cx.override_style);
        cx.override_style = if self.inherit_style {
            style
        } else {
            Style::default()
        };

        if self.fill_with_bg {
            let fill_style = Style {
                bg: style.bg,
                ..Default::default()
            };
            fill_block(cx, to_ratatui_rect(cx.rect()), fill_style);
        }

        render_border(cx, to_ratatui_rect(cx.rect()), &self.border_styles, style);

        self.content.paint(cx)
    }

    fn event(&mut self, cx: &mut EventCx, event: &Event) {
        self.content.event(cx, event)
    }

    fn layout(&mut self, cx: &mut LayoutCx, bc: &BoxConstraints) -> Size {
        let pad = |borders| {
            if self.border_styles.has_borders(borders) {
                1.0
            } else {
                0.0
            }
        };
        let pad_left = pad(Borders::LEFT);
        let pad_right = pad(Borders::RIGHT);
        let pad_top = pad(Borders::TOP);
        let pad_bottom = pad(Borders::BOTTOM);
        let border_padding = Size::new(pad_left + pad_right, pad_top + pad_bottom);
        let content_size = self.content.layout(cx, &bc.shrink(border_padding));

        self.content.set_origin(cx, Point::new(pad_left, pad_top));
        content_size + border_padding
    }

    fn lifecycle(&mut self, cx: &mut super::core::LifeCycleCx, event: &super::LifeCycle) {
        self.content.lifecycle(cx, event);
    }
}
