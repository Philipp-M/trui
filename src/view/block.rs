use crate::widget::{self, ChangeFlags, StyleableWidget};

use super::{BorderKind, BorderStyle, BorderStyles, Borders, Cx, Styleable, View, ViewMarker};
use ratatui::style::{Color, Style};
use xilem_core::MessageResult;

pub struct Block<V> {
    content: V,
    border_styles: BorderStyles,
    style: Style, // base style, merged on top of border style currently (overrides attributes if they are defined in border_style)
    fill_with_bg: bool,
    inherit_style: bool,
}

impl<V> ViewMarker for Block<V> {}

impl<T, A, V> View<T, A> for Block<V>
where
    V: View<T, A>,
    V::Element: 'static,
{
    type State = (V::State, xilem_core::Id);

    type Element = widget::Block;

    fn build(&self, cx: &mut Cx) -> (xilem_core::Id, Self::State, Self::Element) {
        let (id, (state, element)) = cx.with_new_id(|cx| {
            let (child_id, state, element) = self.content.build(cx);
            (
                (state, child_id),
                widget::Block::new(
                    element,
                    self.border_styles.clone(),
                    self.style,
                    self.inherit_style,
                    self.fill_with_bg,
                ),
            )
        });
        (id, state, element)
    }

    fn rebuild(
        &self,
        cx: &mut Cx,
        prev: &Self,
        id: &mut xilem_core::Id,
        (state, child_id): &mut Self::State,
        element: &mut Self::Element,
    ) -> crate::widget::ChangeFlags {
        let mut changeflags = ChangeFlags::empty();
        changeflags |= element.set_border_style(&self.border_styles);
        changeflags |= element.set_style(self.style);
        changeflags |= element.set_inherit_style(self.inherit_style);
        changeflags |= element.set_fill_with_bg(self.fill_with_bg);

        let element = element
            .content
            .downcast_mut()
            .expect("The border content widget changed its type, this should never happen!");

        changeflags
            | cx.with_id(*id, |cx| {
                self.content
                    .rebuild(cx, &prev.content, child_id, state, element)
            })
    }

    fn message(
        &self,
        id_path: &[xilem_core::Id],
        (state, child_id): &mut Self::State,
        message: Box<dyn std::any::Any>,
        app_state: &mut T,
    ) -> MessageResult<A> {
        if let Some((first, rest_path)) = id_path.split_first() {
            if first == child_id {
                self.content.message(rest_path, state, message, app_state)
            } else {
                xilem_core::MessageResult::Stale(message)
            }
        } else {
            xilem_core::MessageResult::Nop
        }
    }
}

crate::impl_event_views!((Block), V, (), (V));

impl<V> Styleable for Block<V> {
    type Output = Self;

    fn fg(mut self, color: Color) -> Self::Output {
        self.style.fg = Some(color);
        self
    }

    fn bg(mut self, color: Color) -> Self::Output {
        self.style.bg = Some(color);
        self
    }

    fn modifier(self, modifier: ratatui::style::Modifier) -> Self::Output {
        let _ = self.style.add_modifier(modifier);
        self
    }

    fn current_style(&self) -> Style {
        self.style
    }

    fn style(mut self, style: Style) -> Self::Output {
        self.style = style;
        self
    }
}

impl<V> Block<V> {
    pub fn inherit_style(mut self, inherit: bool) -> Self {
        self.inherit_style = inherit;
        self
    }

    pub fn fill_with_bg(mut self, fill: bool) -> Self {
        self.fill_with_bg = fill;
        self
    }

    pub fn with_borders(mut self, style: impl Into<BorderStyle>) -> Self {
        self.border_styles.0.push(style.into());
        self
    }

    /// reverse previously applied borders
    pub fn without_borders(mut self, borders: Borders) -> Self {
        self.border_styles.0.push(BorderStyle {
            add_borders: Borders::NONE,
            sub_borders: borders,
            style: Style::default(),
            kind: None,
        });
        self
    }
}

pub fn block<V>(content: V) -> Block<V> {
    Block {
        content,
        border_styles: Default::default(),
        style: Style::default(),
        inherit_style: false,
        fill_with_bg: true,
    }
}

pub fn bordered_block<V>(content: V) -> Block<V> {
    block(content).with_borders(Borders::ALL)
}

// A few From implementations for convenient use with `Block::with_borders()`
// TODO maybe macros may help reducing the boilerplate below...

impl From<()> for BorderStyle {
    fn from(_: ()) -> Self {
        Borders::ALL.into()
    }
}

impl From<Borders> for BorderStyle {
    fn from(add_borders: Borders) -> Self {
        BorderStyle {
            add_borders,
            ..Default::default()
        }
    }
}

// TODO enable all borders?
impl From<Style> for BorderStyle {
    fn from(style: Style) -> Self {
        BorderStyle {
            add_borders: Borders::ALL,
            style,
            ..Default::default()
        }
    }
}

impl From<BorderKind> for BorderStyle {
    fn from(kind: BorderKind) -> Self {
        BorderStyle {
            add_borders: Borders::ALL,
            kind: Some(kind),
            ..Default::default()
        }
    }
}

impl From<(Borders, Style)> for BorderStyle {
    fn from((add_borders, style): (Borders, Style)) -> Self {
        BorderStyle {
            add_borders,
            style,
            ..Default::default()
        }
    }
}

impl From<(Style, Borders)> for BorderStyle {
    fn from((style, add_borders): (Style, Borders)) -> Self {
        BorderStyle {
            add_borders,
            style,
            ..Default::default()
        }
    }
}

impl From<(Borders, BorderKind)> for BorderStyle {
    fn from((borders, kind): (Borders, BorderKind)) -> Self {
        BorderStyle {
            add_borders: borders,
            kind: Some(kind),
            ..Default::default()
        }
    }
}

impl From<(BorderKind, Borders)> for BorderStyle {
    fn from((kind, add_borders): (BorderKind, Borders)) -> Self {
        BorderStyle {
            add_borders,
            kind: Some(kind),
            ..Default::default()
        }
    }
}

impl From<(Style, Borders, BorderKind)> for BorderStyle {
    fn from((style, add_borders, kind): (Style, Borders, BorderKind)) -> Self {
        BorderStyle {
            add_borders,
            sub_borders: Borders::NONE,
            style,
            kind: Some(kind),
        }
    }
}

impl From<(Borders, Style, BorderKind)> for BorderStyle {
    fn from((add_borders, style, kind): (Borders, Style, BorderKind)) -> Self {
        BorderStyle {
            add_borders,
            sub_borders: Borders::NONE,
            style,
            kind: Some(kind),
        }
    }
}

impl From<(Borders, BorderKind, Style)> for BorderStyle {
    fn from((add_borders, kind, style): (Borders, BorderKind, Style)) -> Self {
        BorderStyle {
            add_borders,
            sub_borders: Borders::NONE,
            style,
            kind: Some(kind),
        }
    }
}

impl From<(BorderKind, Borders, Style)> for BorderStyle {
    fn from((kind, add_borders, style): (BorderKind, Borders, Style)) -> Self {
        BorderStyle {
            add_borders,
            sub_borders: Borders::NONE,
            style,
            kind: Some(kind),
        }
    }
}

impl From<(BorderKind, Style, Borders)> for BorderStyle {
    fn from((kind, style, add_borders): (BorderKind, Style, Borders)) -> Self {
        BorderStyle {
            add_borders,
            sub_borders: Borders::NONE,
            style,
            kind: Some(kind),
        }
    }
}

impl From<(Style, BorderKind, Borders)> for BorderStyle {
    fn from((style, kind, add_borders): (Style, BorderKind, Borders)) -> Self {
        BorderStyle {
            add_borders,
            sub_borders: Borders::NONE,
            style,
            kind: Some(kind),
        }
    }
}
