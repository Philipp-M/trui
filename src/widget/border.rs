use super::{core::PaintCx, ChangeFlags, Event, EventCx, Pod, StyleableWidget, Widget};
use crate::view::Borders;
use ratatui::{layout::Rect, style::Style, symbols};
use taffy::tree::NodeId;

pub fn render_border(cx: &mut PaintCx, r: Rect, borders: Borders, style: Style) {
    if r.width == 0 || r.height == 0 {
        return;
    }
    let buf = cx.terminal.current_buffer_mut();

    let mut draw = |x, y, s| {
        if buf.area.x + x < buf.area.width && buf.area.y + y < buf.area.height {
            buf.get_mut(x, y).set_symbol(s).set_style(style);
        }
    };
    if r.width == 1 && r.height == 1 {
        draw(r.x, r.y, symbols::DOT);
        return;
    }
    if r.width > 1 && borders.intersects(Borders::HORIZONTAL) {
        for x in r.x..(r.x + r.width) {
            if borders.contains(Borders::TOP) {
                draw(x, r.y, symbols::line::HORIZONTAL);
            }
            if borders.contains(Borders::BOTTOM) {
                draw(x, r.y + r.height - 1, symbols::line::HORIZONTAL);
            }
        }
    }
    if r.height > 1 && borders.intersects(Borders::VERTICAL) {
        for y in r.y..(r.y + r.height) {
            if borders.contains(Borders::LEFT) {
                draw(r.x, y, symbols::line::VERTICAL);
            }
            if borders.contains(Borders::RIGHT) {
                draw(r.x + r.width - 1, y, symbols::line::VERTICAL);
            }
        }
    }
    if r.width > 1 && r.height > 1 {
        if borders.contains(Borders::LEFT | Borders::TOP) {
            draw(r.x, r.y, symbols::line::ROUNDED_TOP_LEFT);
        }
        if borders.contains(Borders::LEFT | Borders::BOTTOM) {
            draw(r.x, r.y + r.height - 1, symbols::line::ROUNDED_BOTTOM_LEFT);
        }
        if borders.contains(Borders::RIGHT | Borders::BOTTOM) {
            draw(
                r.x + r.width - 1,
                r.y + r.height - 1,
                symbols::line::ROUNDED_BOTTOM_RIGHT,
            );
        }
        if borders.contains(Borders::RIGHT | Borders::TOP) {
            draw(r.x + r.width - 1, r.y, symbols::line::ROUNDED_TOP_RIGHT);
        }
    }
}

pub struct Border {
    pub(crate) content: Pod,
    borders: Borders,
    border_style: Style,
    inherit_style: bool,
}

impl Border {
    pub fn new(
        content: impl Widget + 'static,
        borders: Borders,
        border_style: Style,
        inherit_style: bool,
    ) -> Self {
        Border {
            content: Pod::new(content),
            borders,
            border_style,
            inherit_style,
        }
    }

    pub fn set_borders(&mut self, borders: Borders) -> ChangeFlags {
        if self.borders != borders {
            self.borders = borders;
            ChangeFlags::LAYOUT | ChangeFlags::PAINT
        } else {
            ChangeFlags::empty()
        }
    }

    pub fn set_inherit_style(&mut self, inherit: bool) -> ChangeFlags {
        if self.inherit_style != inherit {
            self.inherit_style = inherit;
            ChangeFlags::LAYOUT | ChangeFlags::PAINT
        } else {
            ChangeFlags::empty()
        }
    }
}

impl StyleableWidget for Border {
    fn set_style(&mut self, style: Style) -> bool {
        let changed = style != self.border_style;
        if changed {
            self.border_style = style;
        }
        changed
    }
}

impl Widget for Border {
    fn paint(&mut self, cx: &mut PaintCx) {
        let style = match cx.override_style {
            Some(style) => style,
            None => self.border_style,
        };

        cx.override_style = if self.inherit_style {
            Some(style)
        } else {
            None
        };

        render_border(cx, cx.rect(), self.borders, style);
        self.content.paint(cx, cx.rect())
    }

    fn layout(&mut self, cx: &mut super::LayoutCx, _prev: NodeId) -> NodeId {
        let pad = |b| {
            taffy::style::LengthPercentage::Length(if self.borders.contains(b) { 1.0 } else { 0.0 })
        };

        let border_style = taffy::style::Style {
            padding: taffy::prelude::Rect {
                left: pad(Borders::LEFT),
                right: pad(Borders::RIGHT),
                top: pad(Borders::TOP),
                bottom: pad(Borders::BOTTOM),
            },
            size: taffy::prelude::Size {
                width: taffy::style::Dimension::Percent(1.0),
                height: taffy::style::Dimension::Percent(1.0),
            },
            ..Default::default()
        };

        // TODO diff children...
        let content = self.content.layout(cx);
        cx.taffy
            .new_with_children(border_style, &[content])
            .unwrap()
    }

    fn event(&mut self, cx: &mut EventCx, event: &Event) {
        self.content.event(cx, event)
    }
}
