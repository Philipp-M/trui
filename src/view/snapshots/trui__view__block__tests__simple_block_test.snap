---
source: src/view/block.rs
expression: buffer
---
Buffer {
    area: Rect { x: 0, y: 0, width: 15, height: 5 },
    content: [
        "some text      ",
        "               ",
        "               ",
        "               ",
        "               ",
    ],
    styles: [
        x: 0, y: 0, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
    ]
}
